x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/io/stream/PlanNamedTypes.java

```diff
@@ -727,6 +728,7 @@ static Enrich readEnrich(PlanStreamInput in) throws IOException {
     static void writeEnrich(PlanStreamOutput out, Enrich enrich) throws IOException {
         out.writeNoSource();
         out.writeLogicalPlanNode(enrich.child());
+        out.writeEnum(enrich.mode());
```

Comment:
I think we need to handle BWC here.
```diff
diff --git a/server/src/main/java/org/elasticsearch/TransportVersions.java b/server/src/main/java/org/elasticsearch/TransportVersions.java
index ce1f6d9ecaad..7d7adc4156de 100644
--- a/server/src/main/java/org/elasticsearch/TransportVersions.java
+++ b/server/src/main/java/org/elasticsearch/TransportVersions.java
@@ -178,6 +178,7 @@ public class TransportVersions {
     public static final TransportVersion SNAPSHOTS_IN_PROGRESS_TRACKING_REMOVING_NODES_ADDED = def(8_566_00_0);
     public static final TransportVersion SMALLER_RELOAD_SECURE_SETTINGS_REQUEST = def(8_567_00_0);
     public static final TransportVersion UPDATE_API_KEY_EXPIRATION_TIME_ADDED = def(8_568_00_0);
+    public static final TransportVersion ESQL_ENRICH_MODE = def(8_569_00_0);
 
     /*
      * STOP! READ THIS FIRST! No, really,
diff --git a/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/io/stream/PlanNamedTypes.java b/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/io/stream/PlanNamedTypes.java
index 7e81207ac919..e8c427870190 100644
--- a/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/io/stream/PlanNamedTypes.java
+++ b/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/io/stream/PlanNamedTypes.java
@@ -7,6 +7,7 @@
 
 package org.elasticsearch.xpack.esql.io.stream;
 
+import org.elasticsearch.TransportVersions;
 import org.elasticsearch.common.TriFunction;
 import org.elasticsearch.common.io.stream.NamedWriteable;
 import org.elasticsearch.common.io.stream.StreamInput;
@@ -717,7 +718,7 @@ public final class PlanNamedTypes {
         return new Enrich(
             in.readSource(),
             in.readLogicalPlanNode(),
-            in.readEnum(Enrich.Mode.class),
+            in.getTransportVersion().onOrAfter(TransportVersions.ESQL_ENRICH_MODE) ? in.readEnum(Enrich.Mode.class) : Enrich.Mode.ANY,
             in.readExpression(),
             in.readNamedExpression(),
             new EnrichPolicyResolution(in.readString(), new EnrichPolicy(in), IndexResolution.valid(readEsIndex(in))),
@@ -728,7 +729,11 @@ public final class PlanNamedTypes {
     static void writeEnrich(PlanStreamOutput out, Enrich enrich) throws IOException {
         out.writeNoSource();
         out.writeLogicalPlanNode(enrich.child());
-        out.writeEnum(enrich.mode());
+        if (out.getTransportVersion().onOrAfter(TransportVersions.ESQL_ENRICH_MODE)) {
+            out.writeEnum(enrich.mode());
+        } else {
+            throw new IllegalArgumentException("cluster is not ready to handle " + enrich.mode().name());
+        }
         out.writeExpression(enrich.policyName());
         out.writeNamedExpression(enrich.matchField());
         out.writeString(enrich.policy().policyName());
```

Reply:
Thanks for confirming, I've added a similar fix.